### PR TITLE
Relaxing element type definitions to HTMLElement

### LINKED ISSUE #648: User event element types are more narrow than what is returned by RTS queries
- `@testing-library/user-event` version:13.1.1

- Testing Framework and version: Jest v26.6.3
  
- DOM Environment: jsdom v16.5.0
  

Relevant code or config

```javascript
userEvent.paste(screen.getByRole('textbox'), 'Ralph');
// or
userEvent.upload(screen.getByLabelText('labelText'), file);
```

What you did:

Compiled using TypeScript.

What happened:

Type error

```
Argument of type 'HTMLElement' is not assignable to parameter of type 'HTMLInputElement | HTMLTextAreaElement'.
  Type 'HTMLElement' is missing the following properties from type 'HTMLTextAreaElement': autocomplete, cols, defaultValue, dirName, and 26 more.ts(2345)
```

Reproduction repository:

N/A

Problem description:

Elements returned by react-testing-library queries are of type `HTMLElement` but user-event functions take more narrow element types like `HTMLInputElement | HTMLLabelElement`. It looks like the type change was made in @testing-library/user-event v13.

Suggested solution:

The type definitions for `userEvent.paste` and `userEvent.upload` should either be relaxed to allow `HTMLElement` or query methods should be added that return the more narrow element types. Type casting should not be required.

### PR DESCRIPTION
**Pull request by GreenGremlin**
Fixes #648

**What**: Relaxing element argument types to match the type returned by react-testing-library queries.

**Why**: 

To allow user-event functions to be used with RTS queries without type casting.

**How**:

I'm narrowing the element type by checking that the passed elements are instances of valid element types. Thus, the check is runtime only, which makes sense since DOM queries are not fully type safe.

**Checklist**:

- [ ] Documentation N/A
- [x] Tests
- [x] Typings
- [x] Ready to be merged

### CODE REVIEW

File: src/__tests__/upload.js [hunk 1]
```diff
@@ -235,3 +235,12 @@ test('input.files implements iterable', () => {
 
   expect(Array.from(eventTargetFiles)).toEqual(files)
 })
+
+test('should give error if we are trying to call upload on an invalid element', () => {
```
Comment (GreenGremlin) [inline src/__tests__/upload.js:239]: This is pretty much just a copy / paste from the existing test for paste.

**Review by ph-fritsche — CHANGES_REQUESTED** [review]
Thanks for taking the time to contribute. :heart:

File: src/paste.ts [hunk 2]
```diff
@@ -23,7 +23,13 @@ function paste(
   }
 
   // TODO: implement for contenteditable
-  if (typeof element.value === 'undefined') {
+  if (
+    !(
+      element instanceof HTMLInputElement ||
+      element instanceof HTMLTextAreaElement
```
[the remark this answers is not in the document]
Comment (GreenGremlin) [reply src/paste.ts:29]: I've updated all `instanceof` checks to use `isElementType`

File: src/paste.ts [hunk 3]
```diff
@@ -23,7 +23,13 @@ function paste(
...
+    ) ||
+    typeof element.value === 'undefined'
```
[the remark this answers is not in the document]
Comment (GreenGremlin) [reply src/paste.ts:31]: I've removed this check

File: src/upload.ts [hunk 4]
```diff
@@ -14,11 +14,20 @@ interface uploadOptions {
...
   {applyAccept = false}: uploadOptions = {},
 ) {
+  if (
+    !(
+      element instanceof HTMLInputElement || element instanceof HTMLLabelElement
+    )
+  ) {
+    throw new TypeError(
+      `the current element is of type ${element.tagName} and does not accept file uploads`,
+    )
```
[the remark this answers is not in the document]
Comment (GreenGremlin) [reply src/upload.ts:29]: Done

File: src/upload.ts [hunk 5]
```diff
@@ -14,11 +14,20 @@ interface uploadOptions {
...
 function upload(
-  element: HTMLInputElement | HTMLLabelElement,
+  element: HTMLElement,
   fileOrFiles: File | File[],
   init?: uploadInit,
   {applyAccept = false}: uploadOptions = {},
 ) {
+  if (
+    !(
+      element instanceof HTMLInputElement || element instanceof HTMLLabelElement
```
[the remark this answers is not in the document]
Comment (GreenGremlin) [reply src/upload.ts:24]: Done

File: src/paste.ts [hunk 6]
```diff
@@ -23,7 +23,13 @@ function paste(
...
+  ) {
     throw new TypeError(
       `the current element is of type ${element.tagName} and doesn't have a valid value`,
```
[the remark this answers is not in the document]
Comment (GreenGremlin) [reply src/paste.ts:34]: I've updated the error message to `the current element is of type ${element.tagName} and does not accept file uploads`

File: src/paste.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ph-fritsche) [reply src/paste.ts:29]: @GreenGremlin You need to push your changes to https://github.com/GreenGremlin/user-event/tree/type-fixes

[the remark this answers is not in the document]
Comment (GreenGremlin) [reply src/paste.ts]: Sorry, they're pushed now.

**Review by ph-fritsche — CHANGES_REQUESTED** [review]
I've opened a PR with code suggestions at https://github.com/GreenGremlin/user-event/pull/1

**Comment by GreenGremlin** [comment]
@ph-fritsche thanks! I've merged your changes.

**Review by ph-fritsche — APPROVED** [review]

**Comment by ph-fritsche** [comment]
@all-contributors add @GreenGremlin code

**Comment by GreenGremlin** [comment]
Thank you!

**Pull request merged**